pandas/core/arrays/datetimes.py

```diff
@@ -477,10 +477,20 @@ def _generate_range(
             arr = arr.astype("M8[ns]", copy=False)
             index = cls._simple_new(arr, freq=None, dtype=dtype)
 
-        if not left_closed and len(index) and index[0] == start:
-            index = index[1:]
-        if not right_closed and len(index) and index[-1] == end:
-            index = index[:-1]
+        # do not remove when one side is inclusive
+        # and removing would leave index empty
+        to_remove_any = not (
+            (left_inclusive or right_inclusive)
+            and len(index)
+            and start == index[0]
+            and start == end
```

Comment:
when is `start != index[0]`?
when is `len(index) == 0` i.e. when will `len(index)` resolve as `False`?
when is `start == end`? 
Do any of these cases imply the others?